From bf7bad24ce141dab5b5acc3ffb98ce5fe4a8e0f9 Mon Sep 17 00:00:00 2001
From: Xie XiuQi <xiexiuqi@huawei.com>
Date: Wed, 21 Oct 2015 15:12:50 +0200
Subject: [PATCH] Fix 'No SMBIOS nor DMI entry point found' on SMBIOS3

address_from_efi may return a SMBIOS or SMBIOS3 format entry
point, so add this condition.
---
 AUTHORS     |  1 +
 CHANGELOG   |  4 ++++
 dmidecode.c | 12 ++++++++++--
 3 files changed, 15 insertions(+), 2 deletions(-)

Index: dmidecode-3.0/AUTHORS
===================================================================
--- dmidecode-3.0.orig/AUTHORS
+++ dmidecode-3.0/AUTHORS
@@ -19,6 +19,7 @@ Jarod Wilson <jarod@redhat.com>
 Anton Arapov <anton@redhat.com>
 Roy Franz <roy.franz@linaro.org>
 Tyler Bell <tyler.bell@hp.com>
+Xie XiuQi <xiexiuqi@huawei.com>
 
 MANY THANKS TO (IN CHRONOLOGICAL ORDER)
 Werner Heuser
Index: dmidecode-3.0/CHANGELOG
===================================================================
--- dmidecode-3.0.orig/CHANGELOG
+++ dmidecode-3.0/CHANGELOG
@@ -1,3 +1,7 @@
+2015-10-21  Xie XiuQi  <xiexiuqi@huawei.com>
+
+	* dmidecode.c: Handle SMBIOS 3.0 entry points on EFI systems.
+
 2015-09-03  Jean Delvare  <jdelvare@suse.de>
 
 	* version.h: Set version to 3.0.
Index: dmidecode-3.0/dmidecode.c
===================================================================
--- dmidecode-3.0.orig/dmidecode.c
+++ dmidecode-3.0/dmidecode.c
@@ -4864,8 +4864,16 @@ int main(int argc, char * const argv[])
 		goto exit_free;
 	}
 
-	if (smbios_decode(buf, opt.devmem, 0))
-		found++;
+	if (memcmp(buf, "_SM3_", 5) == 0)
+	{
+		if (smbios3_decode(buf, opt.devmem, 0))
+			found++;
+	}
+	else if (memcmp(buf, "_SM_", 4) == 0)
+	{
+		if (smbios_decode(buf, opt.devmem, 0))
+			found++;
+	}
 	goto done;
 
 memory_scan:
